accepted/0000-workspaces.md

```diff
@@ -0,0 +1,481 @@
+- Run some npm subcommands in the context of each nested workspace
+
+## Rationale and Alternatives
+
+First and foremost there’s the alternative of leaving the problem set for userland to solve, there’s already the very popular project Lerna that provides some of these features.
+
+Also available is the alternative of supporting only the install (or bootstrap as Lerna names it) aspect of this proposal, following a less feature-rich approach but one that would still enable the basic goal of improving the user experience of managing multiple child packages but from all the feedback collected during the research phase of this RFC, this alternative is much less desirable to the community of maintainers involved.
+
+## Implementation
+
+### 1. Make npm workspace-aware
+
+We're following the lead of Yarn in supporting the `workspaces` `package.json` field which defines a list of paths, each of these paths may be a workspace itself but it also support globs.
+
+`package.json` example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": {
+        "packages": [
+            "packages/*"
+        ]
+    }
+}
+```
+
+`package.json` shorthand example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": [
+        "packages/*"
+    ]
+}
+```
+
+
+### 2. Installing dependencies across child packages
+
+Change `npm install` ([arborist](https://www.npmjs.com/package/@npmcli/arborist)) behavior to make it install every nested package by default once a valid configuration is defined.
+
+Arborist should also be aware of all nested workspaces and correctly link to an internal workspace should it match the required semver version of an expected dependency anywhere in the installing tree. e.g:
+
+```
+// Given this package.json structure:
+├── package.json { "workspaces": ["dep-a", "dep-b"] }
+├── dep-a
+│   └── package.json { "dependencies": { "dep-b": "^1.0.0" } }
+└── dep-b
+    └── package.json { "version": "1.3.1" }
+
+$ npm install
+
+// Results in this symlinking structure:
+├── node_modules
+│   ├── dep-a -> ./dep-a
+│   └── dep-b -> ./dep-b
+├── dep-a
+│   └── node_modules
+│       └── dep-b -> ../../../node_modules/dep-b
+└── dep-b
+```
+
+NOTE: The final implementation of the symlinking structure might differ from the examples laid out here (which are trying to be as illustrative as possible). Arborist might end up resolving all symlinks up to the actual place instead of creating a chain of symlinks.
+
+### 3. Run commands across all child packages
+
+Make npm commands workspace-aware, so that running a command tries to run it within nested packages as long as a workspaces field is defined in `package.json`.
+
+Only a subset of npm commands are to be supported:
```

Comment:
Is the idea to have the npm commands be run against all or subsets of packages? I ask because I'm curious if it's worth it to have commands that are run against just one workspace (ie. `npm workspace package-a run build` or something). Is this what `--filter` is supposed to cover?

Reply:
`--filter` is the way envisioned to reduce the workspaces targeted by the command. By default, all configured workspaces are targeted by the workspace-aware `npm` subcommands.

Reply:
following up on this discussion, we had a quick brainstorm session in which we ended up seeing more benefits adopting [Yarn's approach](https://classic.yarnpkg.com/en/docs/cli/workspace): `npm workspace <workspace_name> <command>` OR `npm ws <workspace_name> <command>` to which I think we could also innovate a bit more and something fancier in the likes of: `npm :<workspace_name> <command>`, so for example running `test` in a specific **workspace** named `a` with these different ideas:

- `npm workspace a test`
- `npm ws a test`
- `npm :a test`

let's follow up with the discussion on [tomorrow's Open RFC call](https://github.com/npm/rfcs/issues/111)

Reply:
...in that call it was decided to completely remove the support to running sub commands from the scope of this RFC and open a follow up instead: #117